### DISCUSSION TITLE
Removing custom codes for hiding specific toolbar buttons?

### DISCUSSION THREAD

**Discussion by cpaschen** [joomengine/Joomla-Component-Builder#935] [Q&A] [answered]
In earlier discussion it was suggested that we try to remove all custom codes inserted via the JCB Custom Codes area - so that it's easier to migrate to JCB-J4.

However, I've not found a way to make changes to the toolbar of a form - to speifically hide the 'save2copy' option (which we need to not be available on several forms).

The only way I've been able to get rid of those is to add a custom code snippet where I comment-out the 'JToolBarHelper::custom(..." line

Is there another way to remove the save2copy button from a toolbar within JCB?

**Comment by dormasta** [comment]
Good evening,

I have the same target to remove specific toolbar buttons. As I also didn't found another way to edit the "HtmlView.php" file I tried Custom Codes.
After watching the video carefully I tried to different was to adapt display(), addToolbar() and addModalToolbar():
1) I "insert<>" and "replace<>" the specific lines I wanted to insert and replace, as it didn't worked I tried
2) to completely replace the 3 functions display(), addToolbar() and addModalToolbar() with one big "replace<>".

The 1) didn't worked because I assumed the blocs were to close to each other creating overlapping code 
The 2) is strange because it creates 2 identical "custom codes", inserts the code twice and and a third time as commented code.

Are there any changes since custom codes in JCB 3 (and the video tutorial) ?
Or is there another way to influence the toolbar?

Cheers

**Comment by lemuelvdm** [comment, marked as answer]
Hi
Overriding the default Joomla toolbar in **Joomla Component Builder (JCB)** has become quite straightforward.

### Where to Manage the Toolbar

In your **Admin View**, you'll find a dedicated **Toolbar** tab. This is the central place where all toolbar behavior is configured.

* **At the top** of this tab, you can add **custom buttons** that extend the default toolbar.
* **At the bottom**, you'll find options to **override the entire default toolbar**, separately for:

  * the **List view**
  * the **Edit view**

Each of these areas allows you to insert **custom PHP code** that will fully replace the default toolbar logic.

### Important: Full Override Behavior

When you add code to override the toolbar, **the entire default toolbar is replaced**.
This includes all standard buttons such as:

* Save
* Save & Close
* Save & New
* Cancel
* etc.

So, that in turn give you the power to **explicitly add back** only the buttons you need to your custom toolbar code.

### Re-using Custom Buttons via Placeholders

If you've already defined **custom buttons** in JCB, you don't need to recreate them manually when overriding the toolbar.

You can re-insert them using the following placeholders:

* `[[[FUNCTION_BUTTONS]]]`
* `[[[CUSTOM_BUTTONS]]]`

What each Placeholder Does:
* `[[[FUNCTION_BUTTONS]]]` -> It adds all Custom Buttons that use `Type = Only Function` to the Toolbar Override in both List View and Item View.
	- Can be used in all toolbars: Admin, Custom Admin, and Site Views.
* `[[[CUSTOM_BUTTONS]]]` -> It adds all Custom Buttons that use `Type = Selection or Default` to the Toolbar Override in both List Views and Item View.
	- Can be used in all toolbars: Admin, Custom Admin, and Site Views.

By placing these placeholders inside your toolbar override code, JCB will automatically inject the relevant buttons back into your custom toolbar. This is especially useful for navigation buttons (e.g. jumping to another view) or batch-style actions.

### About Custom Code Replacement

Regarding your issue with **custom code replacement**:
I strongly recommend going through the [official documentation on **Custom Codes** in JCB.
](https://github.com/joomengine/jcb-documentation/blob/master/english/JCB-Custom-Codes.md)
This feature is powerful, but it can be a bit tricky at first. The documentation explains:

* where to place the code,
* which placeholders are required,
* and how JCB processes and replaces custom code internally.

Taking the time to fully understand this mechanism will help avoid subtle issues and make your implementation much smoother.

---

Hopefully this clarifies how toolbar overrides and custom buttons work in JCB.
If anything is still unclear, feel free to ask.

**Reply by dormasta** [reply]
Hello Lemuel,
Thank you for your fast reply. 
I am absolutely confused, but I don't find this dedicated **Toolbar** tab in the **Admin View**.
Do I need to activate it or is this a JCB 6 feature?
Thanks,
AFT

**Comment by lemuelvdm** [comment]
Hi @dormasta,
My bad - I realize now that you are running the latest stable JCB release. The new Toolbar feature was introduced fairly recently and is currently available only in the latest v5.1.4-beta4. That would explain why you're not seeing it yet.

I would recommend upgrading to this beta version if you'd like access to the latest features and fixes. 

> Yes its still a beta, but we are running ALL our JCB systems on it.

 We are actively working toward the next stable release, and additional beta testing from the Joomla community would be incredibly valuable in helping us deliver an even more robust and polished release.

I have also attached a screenshot of this feature. And yes it is availabe for J5 and J6 but not J4.
<img width="1187" height="632" alt="Screenshot from 2026-01-16 11-29-27" src="https://github.com/user-attachments/assets/2ab7e434-77da-48be-a24a-84a242ba9280" />

Thanks again for your interest and feedback - it's greatly appreciated.

**Reply by dormasta** [reply]
Hello Lemuel,

I installed the beta and I have the Toolbar tab now :-)
May I ask a few questions?
1- 
I already had a custom button including controller and model methods.
The target to adapt the toolbar is to set conditions to the custom buttons depending on the content of the Admin View data.
I set the 'custom button' type to 'only function', I thought this would compile my custom button controller and model method without automatically generating the button in the toolbar above my custom toolbar code. But is adding the custom button above the generated  /***[JCBGUI.admin_view.view_toolbar.xxx.$$$$]***/

2- 
[[[FUNCTION_BUTTONS]]]
[[[CUSTOM_BUTTONS]]]
if I have several custom buttons do I need to add a number to the place holder ?

Thanks 
AFT

**Reply by lemuelvdm** [reply]
Hi, great to hear you've installed the beta and that the **Toolbar tab** is now available

Let me go through your questions one by one.

---

### Question 1 - *Custom button set to "only function" still appearing in the toolbar*

For this one, I'll be honest and keep it simple:
I'm **not fully sure I understand what you're trying to achieve yet** or what exact behavior you're expecting in this case.

Rather than guessing and potentially misleading you, I'd strongly recommend **testing and exploring this functionality in a real, working example** that already makes extensive use of the new toolbar system.

The best reference for this right now is the [**Service Provider Directory**](https://github.com/orgs/joomengine/discussions/1008) component built with JCB. It uses:

* Function buttons
* Custom toolbar placement
* Real-world admin-view logic

Seeing it in action will give far more clarity than a theoretical explanation, especially around how buttons are compiled and rendered.

If, after testing that component, something still doesn't line up with what you're trying to do, feel free to come back with that concrete example and we can drill into it properly.

---

## Question 2 - *Do placeholders need numbering when multiple buttons exist?*

This one I can answer very clearly

### Short answer

**No - you do not add numbers to the placeholders.**

### How placeholders work

These placeholders:

```text
[[[FUNCTION_BUTTONS]]]
[[[CUSTOM_BUTTONS]]]
```

are **collection placeholders**, not single-button placeholders.

That means:

* `[[[FUNCTION_BUTTONS]]]`
-> injects **all buttons with type = function**

* `[[[CUSTOM_BUTTONS]]]`
-> injects **all buttons with type = selection / default**

You can have **one button or many buttons** - you still use the placeholder **once**.

> Wherever you place the placeholder, **all buttons targeted by that placeholder will be rendered there**.

There is:

* No numbering
* No indexing
* No per-button placeholders

This is intentional, to keep toolbars predictable, clean, and easy to maintain.

---

## Recommended Example to Review (Strongly Suggested)

Once again, I highly recommend checking out the **Service Provider Directory** component.

In particular:

* The **"Review Company Updates"** admin view
* It demonstrates the new toolbar and button system exactly as intended

Access to the Blueprint is available to followers of the [**JoomEngine organization**](https://github.com/joomengine), and the process is explained in the relevant [**GitHub discussion**](https://github.com/orgs/joomengine/discussions/1008).

Once you have access:

* Init the component into a test environment
* Inspect the admin views
* You'll immediately see how this new buttons area is meant to function

That should give you the clarity you're looking for.

**Reply by dormasta** [reply]
Hello Lemuel,

from the how to page (https://github.com/orgs/joomengine/discussions/995) it worked fine until 'Initializing the Index' -> 2, then
I get an error 'The repo index failed to load, please try again.'.
For reset, I get the error :
0 Undefined constant "CURLSSLOPT_NO_PARTIALCHAIN" on JROOT/libraries/vendor_jcb/VDM.Joomla.Gitea/src/Utilities/Http/Transport/UnsafeCurl.php:92

Cheers,
AFT


PS: here is the generated code

```
	protected function addToolbar(): void
	{
		$user = $this->getCurrentUser();
		$userId = $user->id;
		$isNew = $this->item->id == 0;
		if ($this->canDo->get('invitation.send_invitation'))

// Following lines is unwanted and automatically added (event with Type/only function) :
		{
			// add myButton button.
			ToolbarHelper::custom('mycomponent.myButton', 'joomla custom-button-myButton', '',           'COM_MYCOMPONENT_MY_BUTTON', false);
		}
// until here is unwanted

/***[JCBGUI.admin_view.view_toolbar.304.$$$$]***/
```

//and then the code content of admin_view -> my component -> Toolbar -> PHP (addToolbar - Admin_view)
